src/libstd/sys_common/backtrace.rs

```diff
@@ -101,50 +101,66 @@ fn filter_frames(frames: &[Frame],
+        "_ZN3std9panicking",
+        "ZN3std9panicking",
+        "std::panicking",
+        "_ZN4core9panicking",
+        "ZN4core9panicking",
+        "core::panicking",
+        "_ZN4core6result13unwrap_failed",
+        "ZN4core6result13unwrap_failed",
+        "rust_begin_unwind",
+        "_ZN4drop",
+        "mingw_set_invalid_parameter_handler",
+    ];
+    static BAD_PREFIXES_BOTTOM: &'static [&'static str] = &[
+        "_ZN3std9panicking",
+        "ZN3std9panicking",
+        "std::panicking",
+        "_ZN4core9panicking",
+        "ZN4core9panicking",
+        "core::panicking",
+        "_ZN3std2rt10lang_start",
+        "ZN3std2rt10lang_start",
+        "__rust_maybe_catch_panic",
+        "_rust_maybe_catch_panic",
+        "__libc_start_main",
+        "__rust_try",
+        "_start",
+        "BaseThreadInitThunk",
+        "__scrt_common_main_seh",
+        "_ZN4drop",
+        "mingw_set_invalid_parameter_handler",
+    ];
+
+    let is_good_frame = |frame: Frame, bad_prefixes: &[&str]| {
+        resolve_symname(frame, |symname| {
             if let Some(mangled_symbol_name) = symname {
-                let magics_begin = [
-                    "_ZN3std3sys3imp9backtrace",
-                    "_ZN3std10sys_common9backtrace",
-                    "_ZN3std9panicking",
-                    "_ZN4core9panicking",
-                    "rust_begin_unwind",
-                    "_ZN4core6result13unwrap_failed",
-                ];
-                if !magics_begin.iter().any(|s| mangled_symbol_name.starts_with(s)) {
-                    skip = true;
+                if !bad_prefixes.iter().any(|s| mangled_symbol_name.starts_with(s)) {
+                    return Ok(())
                 }
             }
-            Ok(())
-        }, context);
+            Err(io::Error::from(io::ErrorKind::Other))
+        }, context).is_ok()
+    };
 
-        if skip {
-            break;
-        }
-    }
+    let skipped_before = frames.iter().position(|frame| {
+        is_good_frame(*frame, BAD_PREFIXES_TOP)
+    }).unwrap_or(frames.len());
+    let skipped_after = frames[skipped_before..].iter().rposition(|frame| {
+        is_good_frame(*frame, BAD_PREFIXES_BOTTOM)
+    }).unwrap_or(frames.len() - skipped_before);
 
-    let mut skipped_after = 0;
-    for (i, frame) in frames.iter().rev().enumerate() {
-        let _ = resolve_symname(*frame, |symname| {
-            if let Some(mangled_symbol_name) = symname {
-                let magics_end = [
-                    "_ZN3std9panicking3try7do_call",
-                    "__rust_maybe_catch_panic",
-                    "__libc_start_main",
-                    "__rust_try",
-                    "_start",
-                ];
-                if magics_end.iter().any(|s| mangled_symbol_name.starts_with(s)) {
-                    skipped_after = i + 1;
-                }
-            }
-            Ok(())
-        }, context);
+    if skipped_before + skipped_after == frames.len() {
```

Comment:
This was true for the old code, but `skipped_after` is counted from `frames[skipped_before..]` now, so the numbers are always in range.